File: backend-dotnet/VaultX.Banking.API/Services/CustomerService.cs
using Microsoft.EntityFrameworkCore;
using VaultX.Banking.API.Data;
using VaultX.Banking.API.DTOs;
using VaultX.Banking.API.Models;

namespace VaultX.Banking.API.Services;

public class CustomerService : ICustomerService
{
    private readonly BankingDbContext _context;

    public CustomerService(BankingDbContext context)
    {
        _context = context;
    }

    public async Task<CustomerResponseDto> RegisterCustomerAsync(CustomerRegistrationDto registrationDto)
    {
        // Check if email already exists
        if (await _context.Customers.AnyAsync(c => c.Email == registrationDto.Email))
        {
            throw new ArgumentException("Email already registered");
        }

        // Check if phone already exists
        if (await _context.Customers.AnyAsync(c => c.Phone == registrationDto.Phone))
        {
            throw new ArgumentException("Phone number already registered");
        }

        var customer = new Customer
        {
            FirstName = registrationDto.FirstName,
            LastName = registrationDto.LastName,
            Email = registrationDto.Email,
            Phone = registrationDto.Phone,
            DateOfBirth = registrationDto.DateOfBirth,
            Password = BCrypt.Net.BCrypt.HashPassword(registrationDto.Password),
            AccountType = Enum.Parse<AccountType>(registrationDto.AccountType, true),
            InitialDeposit = registrationDto.InitialDeposit,
            CurrentBalance = registrationDto.InitialDeposit,
            PanNumber = registrationDto.PanNumber,
            AadharNumber = registrationDto.AadharNumber,
            Occupation = registrationDto.Occupation,
            MonthlyIncome = registrationDto.MonthlyIncome,
            Status = CustomerStatus.DocumentsSubmitted
        };

        _context.Customers.Add(customer);
        await _context.SaveChangesAsync();

        return ConvertToResponseDto(customer);
    }

    public async Task<CustomerLoginResponseDto> LoginCustomerAsync(CustomerLoginDto loginDto)
    {
        var customer = await _context.Customers
            .FirstOrDefaultAsync(c => c.Email == loginDto.EmailOrPhone || c.Phone == loginDto.EmailOrPhone);

        if (customer == null)
        {
            throw new ArgumentException("Customer not found");
        }

        if (!BCrypt.Net.BCrypt.Verify(loginDto.Password, customer.Password))
        {
            throw new ArgumentException("Invalid password");
        }

        return new CustomerLoginResponseDto
        {
            CustomerId = customer.Id,
            Email = customer.Email,
            FirstName = customer.FirstName,
            LastName = customer.LastName,
            Status = customer.Status.ToString(),
            CanPerformOperations = customer.CanPerformOperations
        };
    }

    public async Task<CustomerResponseDto> GetCustomerByIdAsync(int customerId)
    {
        var customer = await _context.Customers.FindAsync(customerId);
        if (customer == null)
        {
            throw new ArgumentException("Customer not found");
        }

        return ConvertToResponseDto(customer);
    }

    public async Task<List<CustomerResponseDto>> GetAllCustomersAsync()
    {
        var customers = await _context.Customers.ToListAsync();
        return customers.Select(ConvertToResponseDto).ToList();
    }

    public async Task<CustomerResponseDto> UpdateCustomerStatusAsync(int customerId, CustomerStatusUpdateDto statusUpdateDto)
    {
        var customer = await _context.Customers.FindAsync(customerId);
        if (customer == null)
        {
            throw new ArgumentException("Customer not found");
        }

        customer.Status = Enum.Parse<CustomerStatus>(statusUpdateDto.Status, true);
        customer.KycCompleted = statusUpdateDto.KycCompleted;
        customer.CanPerformOperations = statusUpdateDto.CanPerformOperations;
        customer.UpdatedAt = DateTime.UtcNow;

        if (statusUpdateDto.Status.Equals("Approved", StringComparison.OrdinalIgnoreCase))
        {
            customer.ApprovalDate = DateOnly.FromDateTime(DateTime.Now);
            customer.AccountNumber = GenerateAccountNumber();
        }

        await _context.SaveChangesAsync();
        return ConvertToResponseDto(customer);
    }

    public async Task<TransactionResponseDto> PerformTransactionAsync(int customerId, TransactionRequestDto transactionDto)
    {
        var customer = await _context.Customers.FindAsync(customerId);
        if (customer == null)
        {
            throw new ArgumentException("Customer not found");
        }

        if (!customer.CanPerformOperations)
        {
            throw new ArgumentException("Customer is not authorized to perform transactions");
        }

        var transactionType = Enum.Parse<TransactionType>(transactionDto.Type, true);

        // Validate transaction
        if (transactionType is TransactionType.Debit or TransactionType.Transfer)
        {
            if (customer.CurrentBalance < transactionDto.Amount)
            {
                throw new ArgumentException("Insufficient balance");
            }
        }

        // Update balance
        decimal newBalance;
        if (transactionType == TransactionType.Credit)
        {
            newBalance = customer.CurrentBalance + transactionDto.Amount;
        }
        else
        {
            newBalance = customer.CurrentBalance - transactionDto.Amount;
        }

        customer.CurrentBalance = newBalance;
        customer.UpdatedAt = DateTime.UtcNow;

        // Create transaction record
        var transaction = new Transaction
        {
            CustomerId = customer.Id,
            TransactionType = transactionType,
            Amount = transactionDto.Amount,
            Description = transactionDto.Description,
            RecipientAccount = transactionDto.RecipientAccount,
            BalanceAfter = newBalance,
            Status = TransactionStatus.Completed,
            ReferenceNumber = "TXN" + DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()
        };

        _context.Transactions.Add(transaction);
        await _context.SaveChangesAsync();

        return ConvertToTransactionResponseDto(transaction);
    }

    public async Task<List<TransactionResponseDto>> GetCustomerTransactionsAsync(int customerId)
    {
        var customer = await _context.Customers.FindAsync(customerId);
        if (customer == null)
        {
            throw new ArgumentException("Customer not found");
        }

        var transactions = await _context.Transactions
            .Where(t => t.CustomerId == customerId)
            .OrderByDescending(t => t.CreatedAt)
            .ToListAsync();

        return transactions.Select(ConvertToTransactionResponseDto).ToList();
    }

    public async Task<BalanceResponseDto> GetCustomerBalanceAsync(int customerId)
    {
        var customer = await _context.Customers.FindAsync(customerId);
        if (customer == null)
        {
            throw new ArgumentException("Customer not found");
        }

        return new BalanceResponseDto
        {
            CurrentBalance = customer.CurrentBalance,
            AccountNumber = customer.AccountNumber,
            AccountType = customer.AccountType.ToString()
        };
    }

    private static CustomerResponseDto ConvertToResponseDto(Customer customer)
    {
        return new CustomerResponseDto
        {
            Id = customer.Id,
            FirstName = customer.FirstName,
            LastName = customer.LastName,
            Email = customer.Email,
            Phone = customer.Phone,
            DateOfBirth = customer.DateOfBirth,
            AccountNumber = customer.AccountNumber,
            AccountType = customer.AccountType.ToString(),
            CurrentBalance = customer.CurrentBalance,
            Status = customer.Status.ToString(),
            KycCompleted = customer.KycCompleted,
            CanPerformOperations = customer.CanPerformOperations,
            SubmittedDate = customer.SubmittedDate,
            KycDate = customer.KycDate,
            ApprovalDate = customer.ApprovalDate
        };
    }

    private static TransactionResponseDto ConvertToTransactionResponseDto(Transaction transaction)
    {
        return new TransactionResponseDto
        {
            Id = transaction.Id,
            TransactionType = transaction.TransactionType.ToString(),
            Amount = transaction.Amount,
            Description = transaction.Description,
            RecipientAccount = transaction.RecipientAccount,
            BalanceAfter = transaction.BalanceAfter,
            Status = transaction.Status.ToString(),
            ReferenceNumber = transaction.ReferenceNumber,
            CreatedAt = transaction.CreatedAt
        };
    }

    private static string GenerateAccountNumber()
    {
        return "VX" + DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
    }
}

File: backend-springboot/src/main/java/com/vaultx/banking/service/CustomerService.java
package com.vaultx.banking.service;

import com.vaultx.banking.dto.*;
import com.vaultx.banking.entity.Customer;
import com.vaultx.banking.entity.Transaction;
import com.vaultx.banking.repository.CustomerRepository;
import com.vaultx.banking.repository.TransactionRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
@Transactional
public class CustomerService {

    @Autowired
    private CustomerRepository customerRepository;

    @Autowired
    private TransactionRepository transactionRepository;

    private final BCryptPasswordEncoder passwordEncoder = new BCryptPasswordEncoder();

    public CustomerResponseDto registerCustomer(CustomerRegistrationDto registrationDto) {
        // Check if email already exists
        if (customerRepository.existsByEmail(registrationDto.getEmail())) {
            throw new RuntimeException("Email already registered");
        }

        // Check if phone already exists
        if (customerRepository.existsByPhone(registrationDto.getPhone())) {
            throw new RuntimeException("Phone number already registered");
        }

        Customer customer = new Customer();
        customer.setFirstName(registrationDto.getFirstName());
        customer.setLastName(registrationDto.getLastName());
        customer.setEmail(registrationDto.getEmail());
        customer.setPhone(registrationDto.getPhone());
        customer.setDateOfBirth(registrationDto.getDateOfBirth());
        customer.setPassword(passwordEncoder.encode(registrationDto.getPassword()));
        customer.setAccountType(Customer.AccountType.valueOf(registrationDto.getAccountType().toUpperCase()));
        customer.setInitialDeposit(registrationDto.getInitialDeposit());
        customer.setCurrentBalance(registrationDto.getInitialDeposit());
        customer.setPanNumber(registrationDto.getPanNumber());
        customer.setAadharNumber(registrationDto.getAadharNumber());
        customer.setOccupation(registrationDto.getOccupation());
        customer.setMonthlyIncome(registrationDto.getMonthlyIncome());
        customer.setStatus(Customer.CustomerStatus.DOCUMENTS_SUBMITTED);

        Customer savedCustomer = customerRepository.save(customer);
        return convertToResponseDto(savedCustomer);
    }

    public CustomerLoginResponseDto loginCustomer(CustomerLoginDto loginDto) {
        Customer customer = customerRepository.findByEmail(loginDto.getEmailOrPhone())
                .orElseGet(() -> customerRepository.findByPhone(loginDto.getEmailOrPhone())
                        .orElseThrow(() -> new RuntimeException("Customer not found")));

        if (!passwordEncoder.matches(loginDto.getPassword(), customer.getPassword())) {
            throw new RuntimeException("Invalid password");
        }

        CustomerLoginResponseDto response = new CustomerLoginResponseDto();
        response.setCustomerId(customer.getId());
        response.setEmail(customer.getEmail());
        response.setFirstName(customer.getFirstName());
        response.setLastName(customer.getLastName());
        response.setStatus(customer.getStatus().name());
        response.setCanPerformOperations(customer.isCanPerformOperations());
        return response;
    }

    public CustomerResponseDto getCustomerById(Long customerId) {
        Customer customer = customerRepository.findById(customerId)
                .orElseThrow(() -> new RuntimeException("Customer not found"));
        return convertToResponseDto(customer);
    }

    public List<CustomerResponseDto> getAllCustomers() {
        return customerRepository.findAll().stream()
                .map(this::convertToResponseDto)
                .collect(Collectors.toList());
    }

    public CustomerResponseDto updateCustomerStatus(Long customerId, CustomerStatusUpdateDto statusUpdateDto) {
        Customer customer = customerRepository.findById(customerId)
                .orElseThrow(() -> new RuntimeException("Customer not found"));

        customer.setStatus(Customer.CustomerStatus.valueOf(statusUpdateDto.getStatus()));
        customer.setKycCompleted(statusUpdateDto.isKycCompleted());
        customer.setCanPerformOperations(statusUpdateDto.isCanPerformOperations());

        if (statusUpdateDto.getStatus().equals("APPROVED")) {
            customer.setApprovalDate(LocalDate.now());
            customer.setAccountNumber(generateAccountNumber());
        }

        Customer savedCustomer = customerRepository.save(customer);
        return convertToResponseDto(savedCustomer);
    }

    public TransactionResponseDto performTransaction(Long customerId, TransactionRequestDto transactionDto) {
        Customer customer = customerRepository.findById(customerId)
                .orElseThrow(() -> new RuntimeException("Customer not found"));

        if (!customer.isCanPerformOperations()) {
            throw new RuntimeException("Customer is not authorized to perform transactions");
        }

        BigDecimal amount = transactionDto.getAmount();
        Transaction.TransactionType type = Transaction.TransactionType.valueOf(transactionDto.getType().toUpperCase());

        // Validate transaction
        if (type == Transaction.TransactionType.DEBIT || type == Transaction.TransactionType.TRANSFER) {
            if (customer.getCurrentBalance().compareTo(amount) < 0) {
                throw new RuntimeException("Insufficient balance");
            }
        }

        // Update balance
        BigDecimal newBalance;
        if (type == Transaction.TransactionType.CREDIT) {
            newBalance = customer.getCurrentBalance().add(amount);
        } else {
            newBalance = customer.getCurrentBalance().subtract(amount);
        }

        customer.setCurrentBalance(newBalance);
        customerRepository.save(customer);

        // Create transaction record
        Transaction transaction = new Transaction();
        transaction.setCustomer(customer);
        transaction.setTransactionType(type);
        transaction.setAmount(amount);
        transaction.setDescription(transactionDto.getDescription());
        transaction.setRecipientAccount(transactionDto.getRecipientAccount());
        transaction.setBalanceAfter(newBalance);
        transaction.setStatus(Transaction.TransactionStatus.COMPLETED);

        Transaction savedTransaction = transactionRepository.save(transaction);
        return convertToTransactionResponseDto(savedTransaction);
    }

    public List<TransactionResponseDto> getCustomerTransactions(Long customerId) {
        Customer customer = customerRepository.findById(customerId)
                .orElseThrow(() -> new RuntimeException("Customer not found"));

        return transactionRepository.findByCustomerOrderByCreatedAtDesc(customer).stream()
                .map(this::convertToTransactionResponseDto)
                .collect(Collectors.toList());
    }

    public BalanceResponseDto getCustomerBalance(Long customerId) {
        Customer customer = customerRepository.findById(customerId)
                .orElseThrow(() -> new RuntimeException("Customer not found"));

        BalanceResponseDto response = new BalanceResponseDto();
        response.setCurrentBalance(customer.getCurrentBalance());
        response.setAccountNumber(customer.getAccountNumber());
        response.setAccountType(customer.getAccountType().name());
        return response;
    }

    private CustomerResponseDto convertToResponseDto(Customer customer) {
        CustomerResponseDto dto = new CustomerResponseDto();
        dto.setId(customer.getId());
        dto.setFirstName(customer.getFirstName());
        dto.setLastName(customer.getLastName());
        dto.setEmail(customer.getEmail());
        dto.setPhone(customer.getPhone());
        dto.setDateOfBirth(customer.getDateOfBirth());
        dto.setAccountNumber(customer.getAccountNumber());
        dto.setAccountType(customer.getAccountType() != null ? customer.getAccountType().name() : null);
        dto.setCurrentBalance(customer.getCurrentBalance());
        dto.setStatus(customer.getStatus().name());
        dto.setKycCompleted(customer.isKycCompleted());
        dto.setCanPerformOperations(customer.isCanPerformOperations());
        dto.setSubmittedDate(customer.getSubmittedDate());
        dto.setKycDate(customer.getKycDate());
        dto.setApprovalDate(customer.getApprovalDate());
        return dto;
    }

    private TransactionResponseDto convertToTransactionResponseDto(Transaction transaction) {
        TransactionResponseDto dto = new TransactionResponseDto();
        dto.setId(transaction.getId());
        dto.setTransactionType(transaction.getTransactionType().name());
        dto.setAmount(transaction.getAmount());
        dto.setDescription(transaction.getDescription());
        dto.setRecipientAccount(transaction.getRecipientAccount());
        dto.setBalanceAfter(transaction.getBalanceAfter());
        dto.setStatus(transaction.getStatus().name());
        dto.setReferenceNumber(transaction.getReferenceNumber());
        dto.setCreatedAt(transaction.getCreatedAt());
        return dto;
    }

    private String generateAccountNumber() {
        return "VX" + System.currentTimeMillis();
    }
}
